candidate_base.go

```diff
@@ -219,16 +218,28 @@ func (c *candidateBase) recvLoop(initializedCh <-chan struct{}) {
 
 	buf := make([]byte, receiveMTU)
 	for {
-		n, srcAddr, err := c.conn.ReadFrom(buf)
-		if err != nil {
-			if !(errors.Is(err, io.EOF) || errors.Is(err, net.ErrClosed)) {
-				a.log.Warnf("Failed to read from candidate %s: %v", c, err)
-			}
+		if err := c.readAndHandlePacket(buf); err != nil {
 			return
 		}
+	}
+}
 
-		c.handleInboundPacket(buf[:n], srcAddr)
+func (c *candidateBase) readAndHandlePacket(buf []byte) error {
+	a := c.agent()
+	muxedConn, isUdpMuxed := c.conn.(*udpMuxedConn)
+	if isUdpMuxed {
+		muxedConn.readMu.Lock()
+		muxedConn.readMu.Unlock()
```

Comment:
Just for understanding this PR: are you missing a `defer` here?

Reply:
I fixed it, but I don't think it is a good idea. Do you have anything other batter way?